Index does not contain the annotation definition: `io.quarkus.security.PermissionsAllowed`

### Describe the bug

Im trying to update some apps from Quarkus 3.34.1 to 3.34.2 and they all fail to build with the following exception:

```
Caused by: java.lang.IllegalArgumentException: Index does not contain the annotation definition: io.quarkus.security.PermissionsAllowed
        at org.jboss.jandex.Index.getAnnotationsWithRepeatable(Index.java:343)
        at org.jboss.jandex.CompositeIndex.getAnnotationsWithRepeatable(CompositeIndex.java:87)
        at org.jboss.jandex.CompositeIndex.getAnnotationsWithRepeatable(CompositeIndex.java:87)
        at org.jboss.jandex.CompositeIndex.getAnnotationsWithRepeatable(CompositeIndex.java:87)
        at io.smallrye.openapi.runtime.scanner.FilteredIndexView.getAnnotationsWithRepeatable(FilteredIndexView.java:335)
        at io.quarkus.security.spi.SecurityTransformerBuildItem$SecurityTransformerImpl.getAnnotations(SecurityTransformerBuildItem.java:488)
        at io.quarkus.security.spi.SecurityTransformerBuildItem$SecurityTransformerImpl.getAnnotationsWithRepeatable(SecurityTransformerBuildItem.java:359)
        at io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor$2.getAnnotationsWithRepeatable(SmallRyeOpenApiProcessor.java:1564)
        at io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor.getPermissionsAllowedMethodReferences(SmallRyeOpenApiProcessor.java:876)
        at io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor.getAuthorizedMethods(SmallRyeOpenApiProcessor.java:824)
        at io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor.getOperlter(SmallRyeOpenApiProcessor.java:805)
        at io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor.lambda$addAutoFilters$10(SmallRyeOpenApiProcessor.java:589)
        at java.base/java.util.HashMap.forEach(HashMap.java:1430)
        at io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor.addAutoFilters(SmallRyeOpenApiProcessor.java:575)
        at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:735)
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:898)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:242)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2651)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2630)
        at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1622)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1589)
        at java.base/java.lang.Thread.run(Thread.java:1474)
        at org.jboss.threads.JBossThread.run(JBossThread.java:501)
```

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

Create an app on https://code.quarkus.io/

Add the following extensions:
- quarkus-rest
- quarkus-smallrye-openapi

Check that it works with `quarkus dev`

Next update it from 3.34.1 to 3.34.2 and you will see the exception in dev mode.

### Output of `uname -a` or `ver`

`Darwin Willem-Jans-MBP 25.4.0 Darwin Kernel Version 25.4.0: Thu Mar 19 19:33:25 PDT 2026; root:xnu-12377.101.15~1/RELEASE_ARM64_T6041 arm64`

### Output of `java -version`

`openjdk 25.0.2 2026-01-20 LTS`

### Quarkus version or git rev

`3.34.2`

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

`Apache Maven 3.9.12 (848fbb4bf2d427b72bdb2471c22fced7ebd9a7a1)`

### Additional information

cc @phillip-kruger

--- a/extensions/smallrye-openapi/deployment/src/main/java/io/quarkus/smallrye/openapi/deployment/SmallRyeOpenApiProcessor.java
+++ b/extensions/smallrye-openapi/deployment/src/main/java/io/quarkus/smallrye/openapi/deployment/SmallRyeOpenApiProcessor.java
@@ -545,6 +545,7 @@ public class SmallRyeOpenApiProcessor {
 
                 return false;
             }
+
         };
 
         return new OpenApiFilteredIndexViewBuildItem(indexView);
@@ -1560,14 +1561,31 @@ public class SmallRyeOpenApiProcessor {
 
             @Override
             public Collection<AnnotationInstance> getAnnotationsWithRepeatable(DotName securityAnnotationName) {
+                // Avoid IndexView#getAnnotationsWithRepeatable as it requires the annotation definition
+                // to be present in the index (to discover the @Repeatable container via reflection).
+                // When MicroProfile filter options like `mp.openapi.scan.*` are in use, the annotation
+                // class may not be visible in the filtered index. Instead, look up both the annotation
+                // and its repeatable container (conventionally named `$List`) directly.
+                DotName containerName = DotName.createSimple(securityAnnotationName.toString() + "$List");
+                Collection<AnnotationInstance> directAnnotations;
+                Collection<AnnotationInstance> containerAnnotations;
                 if (securityTransformer != null) {
-                    return securityTransformer.getAnnotationsWithRepeatable(securityAnnotationName);
+                    directAnnotations = securityTransformer.getAnnotations(securityAnnotationName);
+                    containerAnnotations = securityTransformer.getAnnotations(containerName);
+                } else {
+                    directAnnotations = index.getAnnotations(securityAnnotationName);
+                    containerAnnotations = index.getAnnotations(containerName);
                 }
-                // the annotation definition may not be in the index if the security extension is not on the classpath
-                if (index.getClassByName(securityAnnotationName) == null) {
-                    return Collections.emptyList();
+                List<AnnotationInstance> results = new ArrayList<>(directAnnotations);
+                for (AnnotationInstance container : containerAnnotations) {
+                    AnnotationValue value = container.value();
+                    if (value != null) {
+                        for (AnnotationInstance nested : value.asNestedArray()) {
+                            results.add(AnnotationInstance.create(nested.name(), container.target(), nested.values()));
+                        }
+                    }
                 }
-                return index.getAnnotationsWithRepeatable(securityAnnotationName, index);
+                return results;
             }
 
             @Override

--- a/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/AutoSecurityWithScanPackagesTestCase.java
+++ b/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/AutoSecurityWithScanPackagesTestCase.java
@@ -1,0 +1,40 @@
+package io.quarkus.smallrye.openapi.test.jaxrs;
+
+import org.hamcrest.Matchers;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusExtensionTest;
+import io.restassured.RestAssured;
+
+/**
+ * Verify that setting {@code mp.openapi.scan.packages} does not cause a build failure
+ * when security annotations like {@code @PermissionsAllowed} are not in the filtered index.
+ *
+ * Reproducer for <a href="https://github.com/quarkusio/quarkus/issues/53471">#53471</a>.
+ */
+class AutoSecurityWithScanPackagesTestCase {
+
+    @RegisterExtension
+    static QuarkusExtensionTest runner = new QuarkusExtensionTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClasses(ResourceBean.class, OpenApiResourceSecuredAtMethodLevel.class)
+                    .addAsResource(
+                            new StringAsset("""
+                                    quarkus.smallrye-openapi.security-scheme=jwt
+                                    mp.openapi.scan.packages=io.quarkus.smallrye.openapi.test.jaxrs
+                                    """),
+                            "application.properties"));
+
+    @Test
+    void testOpenApiGeneratedWithScanPackages() {
+        RestAssured.given()
+                .header("Accept", "application/json")
+                .when()
+                .get("/q/openapi")
+                .then()
+                .statusCode(200)
+                .body("paths", Matchers.notNullValue());
+    }
+}

